### PR TITLE
feat: Tabs animate with params set in _changeIndex

### LINKED ISSUE #16892: Transitioning a TabBarView manually doesn't honor the duration and curve
## Steps to Reproduce
Calling `TabController.animateTo` with a duration and curve doesn't affect the behavior of the `TabBarView` animation. Although it's correctly affecting the behavior of the `TabPageSelector` and the `TabBar` animations.

Sample: https://gist.github.com/osama-lionheart/bab886d774f4e87b77104835edffbc49
Video: https://youtu.be/FzRBgflZrR4

## Flutter Doctor
```
[✓] Flutter (Channel master, v0.3.3-pre.10, on Mac OS X 10.13.4 17E199, locale en-US)
[✓] Android toolchain - develop for Android devices (Android SDK 27.0.3)
[✓] iOS toolchain - develop for iOS devices (Xcode 9.3)
[✓] Android Studio
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
[✓] Android Studio (version 3.1)
[✓] VS Code (version 1.22.2)
[✓] Connected devices (1 available)

• No issues found!
```

### PR DESCRIPTION
**Pull request by 6cessfuldev**
This PR modifies the _animateToPage method in TabBarView to accept optional duration and curve parameters set in TabControroller.animateTo method.

fixes #16892 
This PR is based on #122279

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/material/tab_controller.dart [hunk 1]
```diff
@@ -196,19 +196,27 @@ class TabController extends ChangeNotifier {
...
           if (_animationController != null) { // don't notify if we've been disposed
             _indexIsChangingCount -= 1;
+            _indexChangingDuration = null;
+            _indexChangingCurve = null;
             notifyListeners();
           }
         });
     } else {
       _indexIsChangingCount += 1;
+      _indexChangingDuration = animationDuration;
```
Comment (chunhtai) [inline packages/flutter/lib/src/material/tab_controller.dart:214]: do this and next line have any effect?

File: packages/flutter/lib/src/material/tab_controller.dart [hunk 2]
```diff
@@ -243,6 +251,26 @@ class TabController extends ChangeNotifier {
...
   int _indexIsChangingCount = 0;
 
+  /// The Duration used to animate [previousIndex] to [index] as a
+  /// consequence of calling [animateTo].
+  ///
+  /// This value is specifically set during the [animateTo] animation that's
+  /// triggered when the user taps a [TabBar] tab. It is intended to apply the
+  /// animation parameters to the [TabBarView]. It becomes null when [offset]
+  /// is changing as a result of the user dragging (and "flinging") the [TabBarView].
+  Duration? get indexChangingDuration => _indexChangingDuration;
```
Comment (chunhtai) [inline packages/flutter/lib/src/material/tab_controller.dart:261]: Can we avoid exposing transient state in this class?

Comment (chunhtai) [reply packages/flutter/lib/src/material/tab_controller.dart:261]: Imagine someone build code around these transient state, and then we have to make sure these transient behaviors are not to be modified without a breaking change.

**Review by chunhtai — COMMENTED** [review]
Thanks for the PR! I have not looked too deep, just left some initial question, please re-request review to request my attention again.

**Review by chunhtai — COMMENTED** [review]
Hi @6cessfuldev, it looks like the comments are not addressed yet?

File: packages/flutter/lib/src/material/tab_controller.dart (hunk 2 above)
Comment (6cessfuldev) [reply packages/flutter/lib/src/material/tab_controller.dart:257]: @chunhtai  How about make streamcontroller in tabarview and when tabcontroller.animateTo( params ) is executed, params is send to stream. AnimationTick is called in tabbarview and then before pagecontroller.animateToPage is executed, we use tabcontroller.animateTo() 's params from streamcontroller

Comment (chunhtai) [reply packages/flutter/lib/src/material/tab_controller.dart:257]: When the tabbarview _handleTabControllerAnimationTick is called, it actually can access the driving animation, not sure why it abstract away to only use index and guess the duration and curve by itself....

It seems to me we should make sure both PageController and TabController are driven by the same animation. We probably need to do some refactoring around this.

**Review by chunhtai — COMMENTED** [review]
left this comment https://github.com/flutter/flutter/pull/135832/files#r1374894219

File: packages/flutter/lib/src/material/tab_controller.dart (hunk 2 above)
Comment (6cessfuldev) [reply packages/flutter/lib/src/material/tab_controller.dart:257]: @chunhtai You mean, just making a field in tabbar and then using a stream or something to share this field with pageController is a way to end up exposing transient state.

Besides refactoring animation logic is a appropriate way, right?

Sorry  for the late reply and Thank you for your advice and explanation. I learn a lot!

**Comment by 6cessfuldev** [comment]
@chunhtai Thank you for your help!👍

**Pull request closed without merging**